Add support for multiple files when reading properties from CSV

GiudGiud opened the pull request:
closes #25955

I've seen this come up twice in dicussions, and it was a shame not to meet the need

GiudGiud committed:
Rename CSVReader to CSVReaderVPP
XYZReaders are user objects reading a file type.
Many different  systems could want to have a 'csvreader', vpps are not unique to claim the name
refs #25956

GiudGiud committed:
Add support for reading multiple files in PropertyFileReader
- read a new file on every execution
- once all are read, stick with last one
refs #25955

GiudGiud committed:
Add a test for multiple CSV files
refs #25955

GiudGiud committed:
Add documentation for multiple CSV files reading
Improve function from CSV documentation
refs #25955

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1962113/) on e4cc676 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>34a0e1</th>
      <th colspan=4>#26395 e4cc67</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>+0.00%</td>
      <td>86.21%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98357</td>
      <td>98380</td>
      <td>+23</td>
      <td>50</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17264</td>
      <td>17264</td>
      <td>-</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26395/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26395/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26395/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.21% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1962070/) on e4cc676 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26395/site)

This comment will be updated on new commits.

GiudGiud commented:
I have a member attribute that is a reference as needed and yet does not get restarted as expected. I m pretty stumped. @loganharbour or @lindsayad if you can see what I m doing wrong.

Otherwise we dont need to allow recover with everything we do either. I know it does not work now so I put in an error.

lindsayad commented on framework/src/userobjects/PropertyReadFile.C:
```diff
@@ -50,14 +53,23 @@ PropertyReadFile::validParams()
-    _reader(_prop_file_name),
-
+    _prop_file_names(getParam<std::vector<FileName>>("prop_file_name")),
+    _current_file_index(declareRestartableData<unsigned int>(name() + "_file_index", 0)),
```
I assume you're talking about this. Yea I mean this looks fine to me but I'm not the restart expert

GiudGiud replied:
yup this one. I tried int it was the same problem. Is the name() the bad idea? I ll try that

dschwen commented on framework/src/userobjects/PropertyReadFile.C:
```diff
@@ -84,13 +96,33 @@ PropertyReadFile::PropertyReadFile(const InputParameters & parameters)
+    _current_file_index++;
+  }
+  else if (_current_file_index == _prop_file_names.size())
+    mooseDoOnce(
+        mooseInfo("Last file specified has been read. The file will no longer be updated."));
```
`mooseInfo` is already `doOnce`

GiudGiud replied:
```suggestion
    mooseInfo("Last file specified has been read. The file will no longer be updated.");
```

GiudGiud committed:
Address Daniel review

mooseInfo is doonce already

lindsayad commented on framework/src/userobjects/PropertyReadFile.C:
```diff
@@ -84,13 +96,32 @@ PropertyReadFile::PropertyReadFile(const InputParameters & parameters)
+void
+PropertyReadFile::initialize()
+{
+  // Set then read new file, only if we have not reached the last file
+  if (_current_file_index < _prop_file_names.size())
+  {
+    _reader.setFileName(_prop_file_names[_current_file_index]);
+    readData();
+    _current_file_index++;
```
For the very first `initialize` you shouldn't read the data. Or alternatively you shouldn't read the data in the constructor. I tested this on your `piecewise_constant_elem_multiple.i` input after applying this diff
```
diff --git a/framework/src/auxkernels/FunctionAux.C b/framework/src/auxkernels/FunctionAux.C
index 9a5788f0ac..650ba3bd7c 100644
--- a/framework/src/auxkernels/FunctionAux.C
+++ b/framework/src/auxkernels/FunctionAux.C
@@ -9,6 +9,7 @@
 
 #include "FunctionAux.h"
 #include "Function.h"
+#include "PropertyReadFile.h"
 
 registerMooseObject("MooseApp", FunctionAux);
 
@@ -25,6 +26,7 @@ FunctionAux::validParams()
 FunctionAux::FunctionAux(const InputParameters & parameters)
   : AuxKernel(parameters), _func(getFunction("function"))
 {
+  getUserObjectByName<PropertyReadFile>("reader_element");
 }
 
 Real
diff --git a/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i b/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
index 20f22fb908..b5b17bd41e 100644
--- a/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
+++ b/test/tests/functions/piecewise_constant_from_csv/piecewise_constant_elem_multiple.i
@@ -51,6 +51,7 @@
     type = FunctionAux
     variable = 'u'
     function = 'element'
+    execute_on = 'timestep_end'
   []
 []
```
and you get an exodiff. That's because if you apply the true dependency which is that the aux kernel evaluation implicitly depends on the user object execution, then you get a second read of your data before you've evaluated the aux kernel a single time.

GiudGiud replied:
I like reading in the constructor because it get us access to the data really early. 
I ll add some boolean for initalize

GiudGiud replied:
well it s not just that I like it, it s relied upon by a lot of the tests.
I ll find something

lindsayad replied:
Skipping execution on `EXEC_INITIAL` does not solve the problem; in fact you remove `EXEC_INITIAL` as an available execution flag for this class. My patch still leads to an exodiff even with your newest commit. You will need something like an `_initialize_called` boolean

GiudGiud replied:
Umm. I tried that first and hit issues.

I ll try figuring out those issues then

GiudGiud replied:
should be fixed now with force_preaux and no first initialization.

This boolean might not be ideal if the user truly expects a new file read on each execution. I m adding something to the docs

lindsayad commented on test/tests/functions/piecewise_constant_from_csv/tests:
```diff
@@ -50,11 +50,20 @@
+      exodiff = 'piecewise_constant_elem_multiple.e'
+      cli_args = 'Outputs/file_base=piecewise_constant_elem_multiple'
+      detail = 'with multiple CSV files read on every execution of the user object.'
+      # Cannot figure it out so it will error
+      recover = false
```
still unsolved?

GiudGiud replied:
yup. But maybe applying your patch it will resolve it. Let s see

GiudGiud replied:
no it does not help. It s that unsigned int file index that does not get restarted somehow.
@loganharbour do you see anything wrong in how I declare / use `_current_file_index` ?

GiudGiud commented on framework/doc/content/source/functions/PiecewiseConstantFromCSV.md:
```diff
@@ -11,8 +11,9 @@ information from it.
+- is sorted by node-id, in which case, when the function is evaluated at a point, it will locate the node at that point then return the value for that node in the CSV file
 - is sorted by blocks, in which case, when the function is evaluated at a point, it will locate the element containing it then return the value for that element's block in the CSV file
-- defines an interpolation grid, in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
+- defines an interpolation grid (voronoi [!param](/Functions/PiecewiseConstantFromCSV/read_type)), in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
```
```suggestion
- defines an interpolation grid, with the voronoi [!param](/Functions/PiecewiseConstantFromCSV/read_type), in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
```

GiudGiud commented on framework/doc/content/source/vectorpostprocessors/CSVReader.md:
yup.

GiudGiud committed:
Address Alex's review
- keep track of file index within readData
- dont read the file twice at construction then on initial

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Update framework/include/userobjects/PropertyReadFile.h

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/doc/content/source/functions/PiecewiseConstantFromCSV.md:
```diff
@@ -11,8 +11,9 @@ information from it.
+- is sorted by node-id, in which case, when the function is evaluated at a point, it will locate the node at that point then return the value for that node in the CSV file
 - is sorted by blocks, in which case, when the function is evaluated at a point, it will locate the element containing it then return the value for that element's block in the CSV file
-- defines an interpolation grid, in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
+- defines an interpolation grid (voronoi [!param](/Functions/PiecewiseConstantFromCSV/read_type)), in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
```
chanvged

GiudGiud commented on framework/src/userobjects/PropertyReadFile.C:
```diff
@@ -84,13 +96,32 @@ PropertyReadFile::PropertyReadFile(const InputParameters & parameters)
   _bounding_box = MooseUtils::buildBoundingBox(mesh_min, mesh_max);
 
   readData();
+  _current_file_index++;
```
done

lindsayad commented on framework/include/userobjects/PropertyReadFile.h:
```diff
@@ -153,4 +155,7 @@ class PropertyReadFile : public GeneralUserObject
   const unsigned int _nblock;
+
+  /// To keep track of initialization to avoid reading the files twice
+  bool _initialize_called_once;
```
```suggestion
  bool & _initialize_called_once;
```

lindsayad reviewed (changes requested):
After this suggested change is committed, I'd be curious if recover still doesn't work?

GiudGiud commented:
No ive printing the iteration count on a restart and it doesn't get restarted well. It gets reset. 

I ll the fix the Boolean for correctness sake

GiudGiud commented:
... it does work.

lindsayad reviewed (approved):
Happy I get to approve another one before I head out!

GiudGiud commented:
OpenMPI failure unrelated.

GiudGiud commented:
Thanks for the thoughtful review! Getting that recover case working is a huge win thanks to you

hugary1995 commented on framework/src/userobjects/PropertyReadFile.C:
```diff
@@ -50,14 +53,25 @@ PropertyReadFile::validParams()
   params.addParam<bool>(
       "use_zero_based_block_indexing", true, "Are the blocks numbered starting at zero?");
+
+  // Set an execution schedule to what makes sense currently
+  // We do not allow INITIAL because we read the file at construction
+  ExecFlagEnum & exec_enum = params.set<ExecFlagEnum>("execute_on", true);
+  exec_enum.removeAvailableFlags(EXEC_INITIAL, EXEC_FINAL, EXEC_LINEAR, EXEC_NONLINEAR);
+  params.setDocString("execute_on", exec_enum.getDocString());
```
After this change, some users are getting the following error message:
```
Invalid option "INITIAL" in MultiMooseEnum.  Valid options (not case-sensitive) are "NONE POSTCHECK TIMESTEP_END TIMESTEP_BEGIN MULTIAPP_FIXED_POINT_END MULTIAPP_FIXED_POINT_BEGIN CUSTOM"
```

I understand the logic behind disabling INITIAL, FINAL, LINEAR, NONLINEAR because file read is performed at construction time. My question is shouldn't we just disable the `execute_on` parameter entirely?

hugary1995 replied:
maybe tag @GiudGiud

GiudGiud replied:
The first file is read on construction. 

The other files can be read based on the execution schedule which is set by the execute_on flag
